File: training/python/p004_container_with_most_water.py
"""P004 Container With Most Water
Time: O(n), Space: O(1)
"""

def solve(height):
    left, right = 0, len(height) - 1
    best = 0
    while left < right:
        width = right - left
        area = min(height[left], height[right]) * width
        best = max(best, area)
        if height[left] < height[right]:
            left += 1
        else:
            right -= 1
    return best


if __name__ == '__main__':
    print(solve([1, 8, 6, 2, 5, 4, 8, 3, 7]))

File: src/main/java/training/P004ContainerWithMostWater.java
package training;

public class P004ContainerWithMostWater {
    // Time: O(n), Space: O(1)
    public static int solve(int[] height) {
        int left = 0;
        int right = height.length - 1;
        int best = 0;
        while (left < right) {
            int width = right - left;
            int area = Math.min(height[left], height[right]) * width;
            best = Math.max(best, area);
            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }
        return best;
    }

    public static void main(String[] args) {
        System.out.println(solve(new int[] {1, 8, 6, 2, 5, 4, 8, 3, 7}));
    }
}
